sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -105,15 +117,18 @@ def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
                           - interaction_only * nnz ** 2)
 
     # Make the arrays that will form the CSR matrix of the expansion.
-    cdef cnp.ndarray[DATA_T, ndim=1] expanded_data = cnp.ndarray(
-        shape=total_nnz, dtype=data.dtype)
-    cdef cnp.ndarray[INDEX_T, ndim=1] expanded_indices = cnp.ndarray(
-        shape=total_nnz, dtype=indices.dtype)
-    cdef INDEX_T num_rows = indptr.shape[0] - 1
-    cdef cnp.ndarray[INDEX_T, ndim=1] expanded_indptr = cnp.ndarray(
-        shape=num_rows + 1, dtype=indptr.dtype)
-
-    cdef INDEX_T expanded_index = 0, row_starts, row_ends, i, j, k, \
+    cdef DATA_T[:] expanded_data = np.empty(
```

Comment:
Since we're editing that part, maybe we should simplify the logic here by indenting all variable declarations under a single `cdef`:
```python
cdef:
    ...
```